pkg/util/conn_pool.go

```diff
@@ -0,0 +1,198 @@
+}
+
+// Create a new ConnPool instance and start the garbage collector running
+// every @interval.
+func NewConnPool(interval, expiry time.Duration) *ConnPool {
+	cp := ConnPool{
+		interval: interval,
+		expiry:   expiry,
+		lock:     &sync.RWMutex{},
+		conns:    make(map[string]*connEntry),
+	}
+	cp.timer = time.AfterFunc(interval, cp.gc)
+
+	return &cp
+}
+
+// loop through all cp.conns and destroy objects that have not been used for cp.expiry.
+func (cp *ConnPool) gc() {
+	cp.lock.Lock()
+	defer cp.lock.Unlock()
+
+	now := time.Now()
+	for key, ce := range cp.conns {
+		if ce.users == 0 && (now.Sub(ce.lastUsed)) > cp.expiry {
+			ce.destroy()
+			delete(cp.conns, key)
+		}
+	}
+
+	// schedule the next gc() run
+	cp.timer.Reset(cp.interval)
+}
+
+// Stop the garbage collector and destroy all connections in the pool.
+func (cp *ConnPool) Destroy() {
+	cp.timer.Stop()
+	// wait until gc() has finished, in case it is running
+	cp.lock.Lock()
+	defer cp.lock.Unlock()
+
+	for key, ce := range cp.conns {
+		if ce.users != 0 {
+			// should never happen
+			// TODO: return error, or panic()?
+			return
+		}
+
+		ce.destroy()
+		delete(cp.conns, key)
+	}
+}
+
+func (cp *ConnPool) generateUnique(pool, monitors, keyfile string) string {
+	return fmt.Sprintf("%s|%s|%s", pool, monitors, keyfile)
+}
+
+// getExisting returns the existing rados.Conn associated with the unique key.
+//
+// Requires: locked cp.lock because of ce.get()
+func (cp *ConnPool) getConn(unique string) *rados.Conn {
+	ce, exists := cp.conns[unique]
+	if exists {
+		ce.get()
+		return ce.conn
+	}
+	return nil
+}
+
+// Return a rados.Conn for the given arguments. Creates a new rados.Conn in
+// case there is none in the pool. Use the returned unique string to reduce the
+// reference count with ConnPool.Put(unique).
+func (cp *ConnPool) Get(pool, monitors, keyfile string) (*rados.Conn, string, error) {
+	unique := cp.generateUnique(pool, monitors, keyfile)
+
+	cp.lock.RLock()
+	conn := cp.getConn(unique)
+	cp.lock.RUnlock()
+	if conn != nil {
+		return conn, unique, nil
+	}
+
+	// construct and connect a new rados.Conn
+	args := []string{"--pool", pool, "-m", monitors, "--keyfile=" + keyfile}
+	conn, err := rados.NewConn()
+	if err != nil {
+		return nil, "", err
+	}
+	err = conn.ParseCmdLineArgs(args)
+	if err != nil {
+		return nil, "", err
+	}
+
+	err = conn.Connect()
+	if err != nil {
+		return nil, "", err
+	}
+
+	ce := &connEntry{
+		conn:     conn,
+		lastUsed: time.Now(),
+		unique:   unique,
+		users:    1,
+	}
+
+	cp.lock.Lock()
+	defer cp.lock.Unlock()
+	oldConn := cp.getConn(unique)
+	if oldConn != nil {
+		// there was a race, oldConn already exists
+		ce.destroy()
+		return oldConn, unique, nil
+	}
+	// this really is a new connection, add it to the map
+	cp.conns[unique] = ce
+
+	return conn, unique, nil
+}
+
+// Reduce the reference count of the rados.Conn object identified by unique
+// that was returned with ConnPool.Get().
+func (cp *ConnPool) Put(unique string) {
```

Comment:
can we rename this to something else, IMHO `Put` seems like we are storing/adding some data in general terms

Reply:
With reference counting Get()/Put() are commonly used. I can rename it to Ref()/Unref() if you prefer, but the Get() makes that feel a little awkward.

Reply:
Acquire/Release?